src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1426,10 +1426,81 @@ public static unsafe nuint NarrowUtf16ToAscii(char* pUtf16Buffer, byte* pAsciiBu
             goto Finish;
         }
 
-        private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pAsciiBuffer, nuint elementCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool VectorContainsNonAsciiChar(Vector128<ushort> utf16Vector, Vector128<short> asciiMaskForTestZ, Vector128<ushort> asciiMaskForAddSaturate)
         {
-            // This method contains logic optimized for both SSE2 and SSE41. Much of the logic in this method
-            // will be elided by JIT once we determine which specific ISAs we support.
+            if (Sse2.IsSupported)
+            {
+                // See comments in GetIndexOfFirstNonAsciiChar_Sse2 for information about how this works.
+
+                if (Sse41.IsSupported)
+                {
+                    if (!Sse41.TestZ(utf16Vector.AsInt16(), asciiMaskForTestZ))
+                    {
+                        return true;
+                    }
+                }
+                else
+                {
+                    if ((Sse2.MoveMask(Sse2.AddSaturate(utf16Vector, asciiMaskForAddSaturate).AsByte()) & 0b_1010_1010_1010_1010) != 0)
+                    {
+                        return true;
+                    }
+                }
```

Comment:
Just noting that longer term (no expectation to fix in this PR), this could probably just be `return (utf16Vector & asciiMaskForTestZ) != Vector128<ushort>.Zero;`.

`==` and `!=` will automatically use `TestZ` where possible, although we might have a missed opportunity to also fold the `x & y` but that seems like something we could log an issue to fix.

Sse41 is 15 years old at this point and its reasonable to expect most x86/x64 machines will have it so maintaining a separate code path just for pre 2007 machines seems undesirable in the long term.

Reply:
Sure, as the vector API gets more powerful, we can refactor the intrinsics. That might also hide these separate SSE2/SSE41 paths and simplify to remove/unify them in the future.